src/conv/invokers/mlir_impl_gemm.cpp

```diff
@@ -135,33 +165,63 @@ InvokerFactory MakeMlirFwdInvokerFactory(const ConvolutionContext& ctx)
                 primitive_parameters.CastTo<conv::DataInvokeParams>();
             const auto& tensors = forward_invoke_params.tensors;
 
-            void* filter = nullptr;
-            void* input  = nullptr;
-            void* output = nullptr;
-#if MIOPEN_BACKEND_OPENCL
-            clGetMemObjectInfo(tensors.w, CL_MEM_HOST_PTR, sizeof(filter), &filter, nullptr);
-            clGetMemObjectInfo(tensors.in, CL_MEM_HOST_PTR, sizeof(input), &input, nullptr);
-            clGetMemObjectInfo(tensors.out, CL_MEM_HOST_PTR, sizeof(output), &output, nullptr);
-#elif MIOPEN_BACKEND_HIP
-            // NOLINTNEXTLINE (cppcoreguidelines-pro-type-const-cast)
-            filter = const_cast<void*>(tensors.w);
-            // NOLINTNEXTLINE (cppcoreguidelines-pro-type-const-cast)
-            input = const_cast<void*>(tensors.in);
-            // NOLINTNEXTLINE (cppcoreguidelines-pro-type-const-cast)
-            output = const_cast<void*>(tensors.out);
-#endif
+            setMlirConvArgsPtr(tensors, args);
+            handle.Run(kernels[0])(args);
+        };
+    };
+}
+
+InvokerFactory MakeMlirBwdInvokerFactory(const ConvolutionContext& ctx)
+{
+    assert(ctx.direction.IsBackwardData());
+
+    std::vector<size_t> in_dims, in_strides;
+    std::vector<size_t> weights_dims, weights_strides;
+    std::vector<size_t> out_dims, out_strides;
+    permuteDimStridesAllDir(ctx.conv_problem,
+                            in_dims,
+                            in_strides,
+                            weights_dims,
+                            weights_strides,
+                            out_dims,
+                            out_strides);
+    MlirConvArgs args =
+        makeMlirConvArgs(in_dims, in_strides, weights_dims, weights_strides, out_dims, out_strides);
+
+    const auto& conv  = ctx.conv_problem.GetConv();
+    const auto& wDesc = ctx.conv_problem.GetWeights();
+
+    const bool isFilter1Padding0Stride1 = [=]() {
```

Comment:
Hmm... I do not think that you need capturing by-copy here.

Reply:
I don't think it matters in this case. I need to capture `conv` and `wDesc`, both are already const references. I suppose a copy of reference isn't that bad? I may need to make an experiment of some sort to see if the copied reference will invoke copy constructor, that'd be too expensive.

Reply:
> I don't think it matters in this case.

I think it almost always matters. Why: capturing by-reference is almost always cheaper than capturing by-copy. Therefore, there should be reasons for capturing by-copy.

If you do not need copies, then simply use `[&]`. When I see `[=]`, I begin asking myself "why the programmer wants to capture copies?"

Also I am suspecting that both `conv` and `wDesc` objects will be captured by value (but I am not sure).

> copied reference will invoke copy constructor,

Approximately, this is the same as copying the pointer.

Reply:
> I begin asking myself "why the programmer wants to capture copies?

:grinning: Haha, it might be that part of driver development background in me. My intuition has been that capture by copies are safer than reference when you know for sure you don't want the object changed. Unfortunately lambda does not allow const reference so I do copies to be defensive. This discussion is getting to a bit philosophical but I agree at this case by reference is almost for sure cheaper. I'd still curious to find out what actually happens in the original case when capture by copy + reference combination comes together. Will take a look later and update here with the result.

Reply:
> Unfortunately lambda does not allow const reference so I do copies to be defensive.

If you capture a `const` object by-reference, then you have const reference in the lambda.

If the object is not `const`, then consider this: "_A capture with an initializer... makes it possible to capture by const reference, with_ `&cr = std::as_const(x)` _or similar_", see https://en.cppreference.com/w/cpp/language/lambda.

Reply:
[Resolved]